src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java

```diff
@@ -164,35 +189,105 @@ private Object createObjectFromMap(String name) throws CheckstyleException {
     }
 
     /**
-     * Create a new instance of a named class.
-     * @param className the name of the class to instantiate.
-     * @param secondAttempt the set of names to attempt instantiation
-     *                      if usage of the className was not successful.
-     * @return the {@code Object} created by loader or null.
-     * @throws CheckstyleException if the class fails to instantiate.
+     * Generate the map of third party Checkstyle module names to their fully qualified names.
+     * @param loader the class loader used to load Checkstyle package names
+     * @return the map of third party Checkstyle module names to their fully qualified names
      */
-    private Object createObjectWithIgnoringProblems(String className, Set<String> secondAttempt)
-            throws CheckstyleException {
-        Object instance = createObject(className);
-        if (instance == null) {
-            final Iterator<String> ite = secondAttempt.iterator();
-            while (instance == null && ite.hasNext()) {
-                instance = createObject(ite.next());
-            }
+    private Map<String, String> generateThirdPartyNameToFullModuleName(ClassLoader loader) {
+        Map<String, String> returnValue;
+        try {
+            final ClassPath classPath = ClassPath.from(loader);
+            // in packages, the value is something like "com.puppycrawl.tools.checkstyle."
+            // the last "." need to be dropped
+            final List<String> packageNames = packages.stream()
+                    .map(pkg -> pkg.substring(0, pkg.length() - 1))
+                    .collect(Collectors.toList());
+            returnValue = packageNames.stream()
+                    .flatMap(pkg -> classPath.getTopLevelClasses(pkg).stream())
+                    .map(ClassPath.ClassInfo::load)
+                    .filter(PackageObjectFactory::isCheckstyleModule)
+                    .collect(Collectors.toMap(Class::getSimpleName, Class::getCanonicalName));
         }
-        return instance;
+        catch (IOException ignore) {
+            returnValue = new HashMap<>();
+        }
+        return returnValue;
+    }
+
+    /**
+     * Checks whether a class may be considered as a checkstyle module. Checkstyle's modules are
+     * non-abstract classes, and are either checkstyle's checks, file sets, filters, file filters,
+     * or root module.
+     * @param loadedClass class to check.
+     * @return true if the class may be considered as the checkstyle module.
+     */
+    private static boolean isCheckstyleModule(Class<?> loadedClass) {
```

Comment:
My opinion is these should be in their own utility class, and public, so 3rd parties can use it for their own needs.
We are already copying/pasting this same code in sevntu and sonar.
We should remove the methods we copied from test so there is no duplication. I think the utility class should be named something like `CheckstyleReflectionUtility`.

@romani Are you ok with this?